libcontainer/nsenter/nsexec.c

```diff
@@ -704,18 +723,46 @@ void nsexec(void)
 						first_child = child;
 
 						/* Get the init_func pid. */
-						if (read(syncfd, &child, sizeof(child)) != sizeof(child)) {
-							kill(first_child, SIGKILL);
+						if (read
+						    (syncfd, &child,
+						     sizeof(child)) !=
+						    sizeof(child)) {
+							kill(first_child,
+							     SIGKILL);
 							bail("failed to sync with child: read(childpid)");
 						}
 
 						/* Send ACK. */
 						s = SYNC_RECVPID_ACK;
-						if (write(syncfd, &s, sizeof(s)) != sizeof(s)) {
-							kill(first_child, SIGKILL);
+						if (write(syncfd, &s, sizeof(s))
+						    != sizeof(s)) {
+							kill(first_child,
+							     SIGKILL);
 							kill(child, SIGKILL);
 							bail("failed to sync with child: write(SYNC_RECVPID_ACK)");
 						}
+
+						/* Send the init_func pid back to our parent. */
+						/*
+						 *                       * Send the init_func pid and the pid of the first child back to our parent.
+						 *                                               *
+						 *                                                                       * We need to send both back because we can't reap the first child we created (CLONE_PARENT).
+						 *                                                                                               * It becomes the responsibility of our parent to reap the first child.
+						 *                                                                                                                       */
+						len =
+						    snprintf(buf, JSON_MAX,
+							     "{\"pid\": %d, \"pid_first\": %d}\n",
+							     child,
+							     first_child);
+						if (len < 0) {
+							kill(child, SIGKILL);
```

Comment:
len of child and first > (4096 - 20ish?) maybe dump those pids to log? put another way.. how could the length of pid / first be > 12chars? Maybe they are not valid pids?

Reply:
Also this can be done with `dprintf` without the need for an intermediate buffer. I would recommend doing that.